### DISCUSSION TITLE
How does Pydantic provide annotations for __init__ to IDEs?

### DISCUSSION THREAD

**Discussion by berzi** [pydantic/pydantic#5103] [Question]
If I define a Pydantic model and start typing `MyModel(` in PyCharm, it gives me a neat list of parameters based on the fields I defined.

I've given a quick look at BaseModel's definition and its metaclass, but I couldn't get a definitive answer from the code and I sadly don't have enough time to get into the specifics of how Pydantic works. I can tell there must be something to do with `typing_extensions.dataclass_transform()`, but it isn't documented apart from its PEP and a lacklustre and oddly-worded docstring (unless the actual documentation is somewhere I couldn't find).

Moreover, if I make my own class as a bare class with a couple attributes and decorate it with `@dataclass_transform`, all I get from it is that when I type `MyClass(`, the IDE suggests `...: _T` instead of nothing, so I'm clearly missing something.

My use-case would be, if it's easy enough to do, to get some plug-in annotations for Peewee models, which is sadly rather arrogantly untyped, and since fields are defined similarly to Pydantic I wonder if it may be easier this way rather than trying to make the two metaclasses work together. Also, I'm just curious in general.

**Comment by dmontagu** [comment]
I am far from confident, but `@dataclass_transform` was specifically created for the purposes of allowing custom libraries to get dataclass-like behavior in IDEs/etc., so I think that may be a big part of it. Beyond that I don't think we've done anything on our end to help with this (outside of @koxudaxi's pycharm plugin which I use and love, and possibly something analogous on the VSCode side, if it exists?); IDE developers sometimes add custom functionality for improved compatibility with specific libraries, but I'm not sure whether JetBrains/VSCode have done anything along those lines for pydantic.

**Reply by koxudaxi** [reply]
@berzi 
> If I define a Pydantic model and start typing MyModel( in PyCharm, it gives me a neat list of parameters based on the fields I defined.

Did you install Pydantic plugin for PyCharm?
I think PyCharm doesn't support `@dataclass_transform`. 
You don't get the feature in PyCharm now.
https://youtrack.jetbrains.com/issue/PY-54560

VS Code supports `@dataclass_transform` via `Pyrgiht.`
I know the background history. I wrote Pydantic plugin for PyCharm, Because, PyCharm provides plugin features.
But, VSCode(pyright) doesn't provide the feature for a third-party library.
Many people requested to support dataclass like libraries like pydantic, And He created the PEP for `@dataclass_transform` to support the libraries in VS Code.
https://github.com/microsoft/pyright/blob/main/specs/dataclass_transforms.md

> My use-case would be, if it's easy enough to do, to get some plug-in annotations for Peewee models,

I wrote strawberry plugin for Pycharm.
https://github.com/strawberry-graphql/strawberry-pycharm-plugin
We can write the plugin for another library.
However, we have  `@dataclass_transform`. I recommend depending on `@dataclass_transform` if the library doesn't have a special behavior.

@dmontagu 
> I am far from confident, but @dataclass_transform was specifically created for the purposes of allowing custom libraries to get dataclass-like behavior in IDEs/etc.,

Yes, The feature is for IDEs. If I have the time, I may write the plugin or the PR for PyCharm :)

**Reply by berzi** [reply]
Yes, I have Pydantic's plugin. It does work for Pydantic, but I guess the open issue on PyCharm explains why it doesn't work for other classes which use dataclass_transform.

I would like to rely on dataclass_transform, so it seems I'll have to wait until JetBrains acts on the feature request... Thank you.
